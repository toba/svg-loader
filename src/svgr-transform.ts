import SVGO from 'svgo';
import svgr from '@svgr/core';
import { OptimizedSvg } from './svgo-plugin';
import { TransformOptions, BabelFileResult } from '@babel/core';
import metroBabel from 'metro-react-native-babel-transformer';

// interface SVGO {
//    _optimizeOnce: () => string;
// }

export const svgoPlugin: svgr.Plugin = (src, config, state) => {
   if (!config.svgo) {
      return src;
   }
   let out = '';

   const cb = (svg: OptimizedSvg) => {
      out = svg.data;
   };

   // function getInfo(state) {
   //    return state.filePath
   //       ? { input: 'file', path: state.filePath }
   //       : { input: 'string' };
   // }
   const svgo = new SVGO();
   svgo._optimizeOnce(src, null, cb);
   return out;
};

/**
 *
 * @see https://www.smooth-code.com/open-source/svgr/docs/typescript/
 * @see https://github.com/smooth-code/svgr/blob/e3009cb37037e828c3f5360b42ad351fa51222e9/packages/babel-plugin-transform-svg-component/src/index.test.js
 */
const template: svgr.Template = (
   { template },
   _opts,
   { componentName, jsx }
) => {
   const ts = template.smart({ plugins: ['typescript'] });
   return ts.ast`
     import * as React from 'react';
     const ${componentName} = (props: React.SVGProps<SVGSVGElement>) => ${jsx};
     export default ${componentName};
   `;
};

export const svgToJSX = (svg: string) =>
   svgr.sync(svg, {
      icon: false,
      native: true,
      svgo: true,
      plugins: [svgoPlugin],
      template,
      ext: '.ts'
   });

/**
 * @see https://github.com/jamiebuilds/babel-handbook/blob/master/translations/en/plugin-handbook.md
 */
export function transform(
   src: string | { src: string; filename: string; options?: TransformOptions },
   filename: string,
   options?: TransformOptions
): BabelFileResult {
   if (typeof src === 'object') {
      // handle RN >= 0.46
      ({ src, filename, options } = src);
   }

   return filename.endsWith('.svg')
      ? metroBabel.transform({
           src: svgToJSX(src),
           filename,
           options
        })
      : metroBabel.transform({ src, filename, options });
}
