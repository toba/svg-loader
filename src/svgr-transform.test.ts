import '@toba/test';
import path from 'path';
import { readFileText } from '@toba/test';
import { transform, svgToJSX, svgoPlugin } from './svgr-transform';

let svg = '';

beforeAll(async () => {
   svg = await readFileText(
      path.resolve(__dirname, '__mocks__', 'logo-colored.svg')
   );
});

test('uses SVGO to prepare SVG', () => {
   const out = svgoPlugin(svg, { svgo: true }, null);

   ['xmlns:bx', 'xmlns', 'style='].forEach(nope => {
      expect(out.includes(nope)).toBe(false);
   });

   expect(out).toMatchSnapshot();
});

test('converts SVG to a JSX component', () => {
   const jsx = svgToJSX(svg);

   ["React from 'react'", 'SVGSVGElement'].forEach(yep => {
      expect(jsx.includes(yep)).toBe(true);
   });
   expect(jsx).toMatchSnapshot();
});
// https://github.com/smooth-code/svgr/blob/e3009cb37037e828c3f5360b42ad351fa51222e9/packages/babel-plugin-transform-svg-component/src/index.test.js
// test('creates module for SVG file', async () => {
//    const out = transform(svg, 'logo-colored.svg');

//    expect(out).toBeDefined();
//    expect(out.code).toBeDefined();
//    expect(out.code).toMatchSnapshot();
// });
