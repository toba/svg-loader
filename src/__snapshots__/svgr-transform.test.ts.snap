// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converts SVG to a JSX component 1`] = `
"import React from \\"react\\";
import Svg, { Path, Ellipse } from \\"react-native-svg\\";

const SvgComponent = props => <Svg className=\\"frog\\" viewBox=\\"0 0 100 100\\" {...props}><Path fill=\\"#4676b4\\" stroke=\\"#000\\" d=\\"M90.023 73.561c-2.469 3.515-5.145 6.645-7.834 9.258a41.48 41.48 0 0 1 1.269.005c5.563.111 10.101 1.333 10.134 2.73.022.895-1.811 1.645-4.593 2.039 4.591 1.245 7.704 2.856 7.443 4.025-.273 1.229-4.192 1.588-9.326.991 3.003 1.561 4.766 3.005 4.421 3.821-.58 1.37-6.871.458-14.053-2.036-3.115-1.082-5.886-2.284-7.977-3.41-1.905.452-3.569.344-4.864-.413-5.26-3.072-2.424-15.669 6.334-28.136 8.757-12.467 20.121-20.082 25.381-17.01 5.259 3.072 2.423 15.669-6.335 28.136zM6.24 85.554c.034-1.397 4.626-2.619 10.256-2.729a40.3 40.3 0 0 1 1.096-.007c-2.685-2.613-5.351-5.742-7.802-9.257C1.094 61.095-1.606 48.498 3.756 45.426c5.364-3.073 16.76 4.543 25.455 17.01 8.695 12.466 11.396 25.063 6.033 28.135-1.273.729-2.885.857-4.72.46-2.107 1.113-4.876 2.296-7.984 3.363C15.272 96.888 8.907 97.8 8.32 96.43c-.349-.816 1.435-2.26 4.474-3.821-5.196.597-9.161.238-9.437-.991-.264-1.169 2.886-2.78 7.531-4.025-2.816-.394-4.669-1.144-4.648-2.039z\\" className=\\"back-legs\\" /><Ellipse cx={50} cy={65.729} fill=\\"#72a3e2\\" stroke=\\"#000\\" className=\\"torso\\" rx={29.51} ry={25.521} /><Ellipse cx={50} cy={62.176} fill=\\"#b4cff2\\" stroke=\\"#000\\" className=\\"front-legs\\" rx={20.093} ry={22.849} /><Path fill=\\"#b4cff2\\" stroke=\\"#000\\" d=\\"M73.576 98.11c-.086.84-2.428 1.256-5.231.929a14.85 14.85 0 0 1-2.744-.567c-.449.707-1.26 1.184-2.122 1.168-.924-.017-1.637-.597-1.842-1.401-1.096.445-2.556.854-4.178 1.138-3.564.622-6.564.406-6.7-.485-.116-.755 1.867-1.753 4.648-2.419-2.155-.741-3.584-1.708-3.445-2.45.185-.989 3.081-1.205 6.469-.483.333.071.659.149.975.232-.441-.869-.516-1.685-.117-2.126.55-.607 1.818-.298 2.99.672 3.544-6.309 10.393-18.793 10.171-20.656-.302-2.529-15.543-9.995-20.245-7.928 0 0 13.531-16.014 14.204-14.951 4.276 6.758 9.138 12.788 14.921 18.085 3.647 3.341-9.8 24.001-13.195 29.08.171.015.344.033.519.053 2.804.325 5.007 1.27 4.922 2.109zM47.795 63.743c-4.702-2.068-19.943 5.403-20.245 7.934-.222 1.861 6.611 14.325 10.159 20.646 1.174-.978 2.449-1.292 3.001-.683.399.441.325 1.257-.117 2.127.317-.083.642-.161.976-.232 3.388-.722 6.284-.506 6.469.483.139.743-1.29 1.711-3.444 2.453 2.78.666 4.762 1.665 4.647 2.421-.136.891-3.136 1.108-6.7.484-1.622-.283-3.082-.692-4.178-1.137-.205.804-.918 1.384-1.842 1.401-.861.016-1.673-.463-2.122-1.17-.781.249-1.723.45-2.744.569-2.803.327-5.145-.091-5.231-.93-.085-.84 2.118-1.785 4.922-2.112.168-.019.334-.036.498-.05-3.429-5.135-16.814-25.732-13.174-29.068 5.783-5.3 10.645-11.333 14.921-18.096.673-1.063 14.204 14.96 14.204 14.96z\\" className=\\"front-legs\\" /><Path fill=\\"#b4cff2\\" d=\\"M33.941 49.488h32.01L49.946 81.919z\\" className=\\"front-legs\\" /><Ellipse cx={50} cy={26.946} fill=\\"#72a3e2\\" stroke=\\"#000\\" className=\\"head\\" rx={40} ry={25} /><Ellipse cx={25} cy={16} fill=\\"#72a3e2\\" stroke=\\"#000\\" className=\\"head\\" rx={15} ry={15} /><Ellipse cx={15.526} cy={27.05} fill=\\"#4676b4\\" className=\\"cheek\\" rx={3.707} ry={6.564} /><Ellipse cx={84.475} cy={27.05} fill=\\"#4676b4\\" className=\\"cheek\\" rx={3.707} ry={6.564} /><Ellipse cx={75} cy={16} fill=\\"#72a3e2\\" stroke=\\"#000\\" className=\\"head\\" rx={15} ry={15} /><Path fill=\\"#72a3e2\\" d=\\"M21.105 11.948l58.656.363v19.653L21.105 31.6z\\" className=\\"head\\" /><Ellipse cx={22.008} cy={16} fill=\\"#fff\\" stroke=\\"#000\\" className=\\"eyes\\" rx={15} ry={15} /><Ellipse cx={78.326} cy={16} fill=\\"#fff\\" stroke=\\"#000\\" className=\\"eyes\\" rx={15} ry={15} /><Path d=\\"M71.654 14.914l3.198-.689a4.006 4.006 0 0 1 2.954-2.634 4.008 4.008 0 0 1 3.778 1.186l3.197-.689a2.054 2.054 0 1 1 .865 4.017l-3.197.688a4.008 4.008 0 0 1-6.733 1.449l-3.198.689a2.055 2.055 0 1 1-.864-4.017zM29.027 14.914a2.054 2.054 0 1 1-.864 4.017l-3.198-.689a4.008 4.008 0 0 1-6.733-1.449l-3.197-.688a2.054 2.054 0 1 1 .865-4.017l3.197.689a4.008 4.008 0 0 1 3.778-1.186 4.006 4.006 0 0 1 2.954 2.634z\\" className=\\"pupil\\" /><Ellipse cx={71.09} cy={36.577} className=\\"nostril\\" rx={1.77} ry={1.77} transform=\\"matrix(1 0 0 1 -24.188 -14.047)\\" /><Ellipse cx={71.09} cy={36.577} className=\\"nostril\\" rx={1.77} ry={1.77} transform=\\"matrix(1 0 0 1 -17.994 -14.047)\\" /><Path fill=\\"#f90\\" stroke=\\"#000\\" d=\\"M39.951 6.139h20.022L49.962 17.185z\\" className=\\"forehead\\" /><Path fill=\\"#b4cff2\\" stroke=\\"#000\\" d=\\"M13.413 37.067c8.934-2.855 22.014-4.655 36.587-4.655 14.573 0 27.653 1.8 36.587 4.655C80.37 45.832 66.327 51.946 50 51.946s-30.37-6.114-36.587-14.879z\\" className=\\"chin\\" /><Path d=\\"M69.952 33.61c-4.017 1.645-11.17 2.739-19.325 2.739-8.351 0-15.651-1.147-19.608-2.858 5.834-.695 12.248-1.079 18.981-1.079 7.109 0 13.862.428 19.952 1.198z\\" className=\\"mouth\\" /><Path fill=\\"#d8d8d8\\" stroke=\\"#000\\" d=\\"M81.714 67.143l6.229-8.646M18.34 67.446L12.111 58.8M90.266 87.875l-5.578-1.023M87.616 92.764l-6.694-2.231M11.586 93.009l6.694-2.231M9.617 87.844l5.578-1.023\\" /><Path stroke=\\"#000\\" d=\\"M57.629 96.05l-3.02.542M41.542 95.933l3.02.542\\" /><Path fill=\\"none\\" stroke=\\"#000\\" d=\\"M4.497 57.087s-1.302-2.435-1.085-5.499c.203-2.876 1.24-3.331 1.24-3.331M81.33 54.841s1.549-1.639 3.331-3.099c1.787-1.464 3.563-2.246 3.563-2.246\\" /></Svg>;

export default SvgComponent;"
`;

exports[`uses SVGO to prepare SVG 1`] = `"<svg class=\\"frog\\" viewBox=\\"0 0 100 100\\"><path fill=\\"#4676b4\\" stroke=\\"#000\\" d=\\"M90.023 73.561c-2.469 3.515-5.145 6.645-7.834 9.258a41.48 41.48 0 0 1 1.269.005c5.563.111 10.101 1.333 10.134 2.73.022.895-1.811 1.645-4.593 2.039 4.591 1.245 7.704 2.856 7.443 4.025-.273 1.229-4.192 1.588-9.326.991 3.003 1.561 4.766 3.005 4.421 3.821-.58 1.37-6.871.458-14.053-2.036-3.115-1.082-5.886-2.284-7.977-3.41-1.905.452-3.569.344-4.864-.413-5.26-3.072-2.424-15.669 6.334-28.136 8.757-12.467 20.121-20.082 25.381-17.01 5.259 3.072 2.423 15.669-6.335 28.136zM6.24 85.554c.034-1.397 4.626-2.619 10.256-2.729a40.3 40.3 0 0 1 1.096-.007c-2.685-2.613-5.351-5.742-7.802-9.257C1.094 61.095-1.606 48.498 3.756 45.426c5.364-3.073 16.76 4.543 25.455 17.01 8.695 12.466 11.396 25.063 6.033 28.135-1.273.729-2.885.857-4.72.46-2.107 1.113-4.876 2.296-7.984 3.363C15.272 96.888 8.907 97.8 8.32 96.43c-.349-.816 1.435-2.26 4.474-3.821-5.196.597-9.161.238-9.437-.991-.264-1.169 2.886-2.78 7.531-4.025-2.816-.394-4.669-1.144-4.648-2.039z\\" class=\\"back-legs\\"/><ellipse cx=\\"50\\" cy=\\"65.729\\" fill=\\"#72a3e2\\" stroke=\\"#000\\" class=\\"torso\\" rx=\\"29.51\\" ry=\\"25.521\\"/><ellipse cx=\\"50\\" cy=\\"62.176\\" fill=\\"#b4cff2\\" stroke=\\"#000\\" class=\\"front-legs\\" rx=\\"20.093\\" ry=\\"22.849\\"/><path fill=\\"#b4cff2\\" stroke=\\"#000\\" d=\\"M73.576 98.11c-.086.84-2.428 1.256-5.231.929a14.85 14.85 0 0 1-2.744-.567c-.449.707-1.26 1.184-2.122 1.168-.924-.017-1.637-.597-1.842-1.401-1.096.445-2.556.854-4.178 1.138-3.564.622-6.564.406-6.7-.485-.116-.755 1.867-1.753 4.648-2.419-2.155-.741-3.584-1.708-3.445-2.45.185-.989 3.081-1.205 6.469-.483.333.071.659.149.975.232-.441-.869-.516-1.685-.117-2.126.55-.607 1.818-.298 2.99.672 3.544-6.309 10.393-18.793 10.171-20.656-.302-2.529-15.543-9.995-20.245-7.928 0 0 13.531-16.014 14.204-14.951 4.276 6.758 9.138 12.788 14.921 18.085 3.647 3.341-9.8 24.001-13.195 29.08.171.015.344.033.519.053 2.804.325 5.007 1.27 4.922 2.109zM47.795 63.743c-4.702-2.068-19.943 5.403-20.245 7.934-.222 1.861 6.611 14.325 10.159 20.646 1.174-.978 2.449-1.292 3.001-.683.399.441.325 1.257-.117 2.127.317-.083.642-.161.976-.232 3.388-.722 6.284-.506 6.469.483.139.743-1.29 1.711-3.444 2.453 2.78.666 4.762 1.665 4.647 2.421-.136.891-3.136 1.108-6.7.484-1.622-.283-3.082-.692-4.178-1.137-.205.804-.918 1.384-1.842 1.401-.861.016-1.673-.463-2.122-1.17-.781.249-1.723.45-2.744.569-2.803.327-5.145-.091-5.231-.93-.085-.84 2.118-1.785 4.922-2.112.168-.019.334-.036.498-.05-3.429-5.135-16.814-25.732-13.174-29.068 5.783-5.3 10.645-11.333 14.921-18.096.673-1.063 14.204 14.96 14.204 14.96z\\" class=\\"front-legs\\"/><path fill=\\"#b4cff2\\" d=\\"M33.941 49.488h32.01L49.946 81.919z\\" class=\\"front-legs\\"/><ellipse cx=\\"50\\" cy=\\"26.946\\" fill=\\"#72a3e2\\" stroke=\\"#000\\" class=\\"head\\" rx=\\"40\\" ry=\\"25\\"/><ellipse cx=\\"25\\" cy=\\"16\\" fill=\\"#72a3e2\\" stroke=\\"#000\\" class=\\"head\\" rx=\\"15\\" ry=\\"15\\"/><ellipse cx=\\"15.526\\" cy=\\"27.05\\" fill=\\"#4676b4\\" class=\\"cheek\\" rx=\\"3.707\\" ry=\\"6.564\\"/><ellipse cx=\\"84.475\\" cy=\\"27.05\\" fill=\\"#4676b4\\" class=\\"cheek\\" rx=\\"3.707\\" ry=\\"6.564\\"/><ellipse cx=\\"75\\" cy=\\"16\\" fill=\\"#72a3e2\\" stroke=\\"#000\\" class=\\"head\\" rx=\\"15\\" ry=\\"15\\"/><path fill=\\"#72a3e2\\" d=\\"M21.105 11.948l58.656.363v19.653L21.105 31.6z\\" class=\\"head\\"/><ellipse cx=\\"22.008\\" cy=\\"16\\" fill=\\"#fff\\" stroke=\\"#000\\" class=\\"eyes\\" rx=\\"15\\" ry=\\"15\\"/><ellipse cx=\\"78.326\\" cy=\\"16\\" fill=\\"#fff\\" stroke=\\"#000\\" class=\\"eyes\\" rx=\\"15\\" ry=\\"15\\"/><path d=\\"M71.654 14.914l3.198-.689a4.006 4.006 0 0 1 2.954-2.634 4.008 4.008 0 0 1 3.778 1.186l3.197-.689a2.054 2.054 0 1 1 .865 4.017l-3.197.688a4.008 4.008 0 0 1-6.733 1.449l-3.198.689a2.055 2.055 0 1 1-.864-4.017zM29.027 14.914a2.054 2.054 0 1 1-.864 4.017l-3.198-.689a4.008 4.008 0 0 1-6.733-1.449l-3.197-.688a2.054 2.054 0 1 1 .865-4.017l3.197.689a4.008 4.008 0 0 1 3.778-1.186 4.006 4.006 0 0 1 2.954 2.634z\\" class=\\"pupil\\"/><ellipse cx=\\"71.09\\" cy=\\"36.577\\" class=\\"nostril\\" rx=\\"1.77\\" ry=\\"1.77\\" transform=\\"matrix(1 0 0 1 -24.188 -14.047)\\"/><ellipse cx=\\"71.09\\" cy=\\"36.577\\" class=\\"nostril\\" rx=\\"1.77\\" ry=\\"1.77\\" transform=\\"matrix(1 0 0 1 -17.994 -14.047)\\"/><path fill=\\"#f90\\" stroke=\\"#000\\" d=\\"M39.951 6.139h20.022L49.962 17.185z\\" class=\\"forehead\\"/><path fill=\\"#b4cff2\\" stroke=\\"#000\\" d=\\"M13.413 37.067c8.934-2.855 22.014-4.655 36.587-4.655 14.573 0 27.653 1.8 36.587 4.655C80.37 45.832 66.327 51.946 50 51.946s-30.37-6.114-36.587-14.879z\\" class=\\"chin\\"/><path d=\\"M69.952 33.61c-4.017 1.645-11.17 2.739-19.325 2.739-8.351 0-15.651-1.147-19.608-2.858 5.834-.695 12.248-1.079 18.981-1.079 7.109 0 13.862.428 19.952 1.198z\\" class=\\"mouth\\"/><path fill=\\"#d8d8d8\\" stroke=\\"#000\\" d=\\"M81.714 67.143l6.229-8.646M18.34 67.446L12.111 58.8M90.266 87.875l-5.578-1.023M87.616 92.764l-6.694-2.231M11.586 93.009l6.694-2.231M9.617 87.844l5.578-1.023\\"/><path stroke=\\"#000\\" d=\\"M57.629 96.05l-3.02.542M41.542 95.933l3.02.542\\"/><path fill=\\"none\\" stroke=\\"#000\\" d=\\"M4.497 57.087s-1.302-2.435-1.085-5.499c.203-2.876 1.24-3.331 1.24-3.331M81.33 54.841s1.549-1.639 3.331-3.099c1.787-1.464 3.563-2.246 3.563-2.246\\"/></svg>"`;
